### DISCUSSION TITLE
Does PyG have hard negative sampling?

### DISCUSSION THREAD

**Discussion by Xurorx** [pyg-team/pytorch_geometric#7747] [Q&A]
Hello PyG team,

I wonder if PyG provides a method to sample hard negative sampling for edge prediction task? I feel like random negative sampling lowers the model performance and I wish there is a function to do the hard negative samples (exculded postive edges).

**Comment by EdisonLeeeee** [comment]
No, as far as I know. What is the definition of "hard negative samples" in a graph? Do you have any references?

**Reply by Xurorx** [reply]
Hello, in this paper there is the definition of "hard negative samples". https://arxiv.org/pdf/1806.01973.pdf .

"Hard negative samples" means items that are somewhat related to the query item q, but not as related
as the positive item i. We call these “hard negative items”. They
are generated by ranking items in a graph according to their Personalized PageRank scores with respect to query item q. Items
ranked at 2000-5000 are randomly sampled as hard negative items."

**Reply by EdisonLeeeee** [reply]
Thank you for your information. This is not yet implemented in PyG, but I can add it to my list :)
